From 416a66de9d428a11776331926c23a61188b8fc16 Mon Sep 17 00:00:00 2001
From: Ander Conselvan de Oliveira <ander@localhost.(none)>
Date: Thu, 12 Mar 2009 09:45:57 -0300
Subject: [PATCH 908/911] Xorg: add an extra module path

If the extra module path is not an empty string (the default value),
module path will be searched first in the extra module path and then in
the default module path. This should simplify the alternatives system
used on Mandriva's fglrx package.
---
Index: xorg-server/configure.ac
===================================================================
--- xorg-server.orig/configure.ac	2011-04-09 10:38:10.000000000 +0300
+++ xorg-server/configure.ac	2011-04-09 10:42:59.000000000 +0300
@@ -489,6 +489,10 @@
 				  [Directory where modules are installed (default: $libdir/xorg/modules)]),
 				[ moduledir="$withval" ],
 				[ moduledir="${libdir}/xorg/modules" ])
+AC_ARG_WITH(extra-module-dir,AS_HELP_STRING([--with-extra-module-dir=DIR],
+				  [Extra module directory to search for modules before the default one (default: empty)]),
+				[ extra_moduledir="$withval" ],
+				[ extra_moduledir="" ])
 AC_ARG_WITH(log-dir,         AS_HELP_STRING([--with-log-dir=DIR],
 				  [Directory where log files are kept (default: $localstatedir/log)]),
 				[ logdir="$withval" ],
@@ -1815,6 +1819,7 @@
 	AC_DEFINE_DIR(XF86CONFIGFILE, XF86CONFIGFILE, [Name of configuration file])
 	AC_DEFINE_DIR(__XCONFIGDIR__, XF86CONFIGDIR, [Name of configuration directory])
 	AC_DEFINE_DIR(DEFAULT_MODULE_PATH, moduledir, [Default module search path])
+	AC_DEFINE_DIR(EXTRA_MODULE_PATH, extra_moduledir, [Extra module search path, searched before the default one])
 	AC_DEFINE_DIR(DEFAULT_LIBRARY_PATH, libdir, [Default library install path])
 	AC_DEFINE_DIR(DEFAULT_LOGPREFIX, LOGPREFIX, [Default log location])
 	AC_DEFINE_UNQUOTED(__VENDORDWEBSUPPORT__, ["$VENDOR_WEB"], [Vendor web address for support])
Index: xorg-server/hw/xfree86/common/xf86Config.c
===================================================================
--- xorg-server.orig/hw/xfree86/common/xf86Config.c	2011-04-09 10:36:56.000000000 +0300
+++ xorg-server/hw/xfree86/common/xf86Config.c	2011-04-09 10:38:10.000000000 +0300
@@ -628,11 +628,21 @@
   
   /* ModulePath */
 
-  if (fileconf) {
-    if (xf86ModPathFrom != X_CMDLINE && fileconf->file_modulepath) {
+  if (xf86ModPathFrom != X_CMDLINE) {
+    if (fileconf && fileconf->file_modulepath) {
       xf86ModulePath = fileconf->file_modulepath;
       xf86ModPathFrom = X_CONFIG;
     }
+    else if (strcmp(xf86ExtraModulePath, "") != 0) {
+      char *newpath = malloc(strlen(xf86ExtraModulePath)
+			     + strlen(xf86ModulePath)
+			     + 2);
+      strcpy(newpath, xf86ExtraModulePath);
+      strcat(newpath, ",");
+      strcat(newpath, xf86ModulePath);
+
+      xf86ModulePath = newpath;
+    }
   }
 
   xf86Msg(xf86ModPathFrom, "ModulePath set to \"%s\"\n", xf86ModulePath);
Index: xorg-server/hw/xfree86/common/xf86Globals.c
===================================================================
--- xorg-server.orig/hw/xfree86/common/xf86Globals.c	2011-04-09 10:36:56.000000000 +0300
+++ xorg-server/hw/xfree86/common/xf86Globals.c	2011-04-09 10:38:10.000000000 +0300
@@ -139,6 +139,7 @@
 const char *xf86ConfigFile = NULL;
 const char *xf86ConfigDir = NULL;
 const char *xf86ModulePath = DEFAULT_MODULE_PATH;
+const char *xf86ExtraModulePath = EXTRA_MODULE_PATH;
 MessageType xf86ModPathFrom = X_DEFAULT;
 const char *xf86LogFile = DEFAULT_LOGPREFIX;
 MessageType xf86LogFileFrom = X_DEFAULT;
Index: xorg-server/hw/xfree86/common/xf86Priv.h
===================================================================
--- xorg-server.orig/hw/xfree86/common/xf86Priv.h	2011-04-09 10:36:56.000000000 +0300
+++ xorg-server/hw/xfree86/common/xf86Priv.h	2011-04-09 10:38:10.000000000 +0300
@@ -75,6 +75,7 @@
 
 extern _X_EXPORT  xf86InfoRec xf86Info;
 extern _X_EXPORT  const char *xf86ModulePath;
+extern _X_EXPORT  const char *xf86ExtraModulePath;
 extern _X_EXPORT  MessageType xf86ModPathFrom;
 extern _X_EXPORT  const char *xf86LogFile;
 extern _X_EXPORT  MessageType xf86LogFileFrom;
Index: xorg-server/include/xorg-config.h.in
===================================================================
--- xorg-server.orig/include/xorg-config.h.in	2011-04-09 10:36:56.000000000 +0300
+++ xorg-server/include/xorg-config.h.in	2011-04-09 10:38:10.000000000 +0300
@@ -42,6 +42,9 @@
 /* Path to loadable modules. */
 #undef DEFAULT_MODULE_PATH
 
+/* Path to extra loadable modules. */
+#undef EXTRA_MODULE_PATH
+
 /* Path to installed libraries. */
 #undef DEFAULT_LIBRARY_PATH
 
